core/src/main/kotlin/links/DRI.kt

```diff
@@ -167,8 +161,66 @@ object StarProjection : TypeReference() {
     override fun toString() = "*"
 }
 
-private operator fun <T> List<T>.component6(): T = get(5)
-
 private val KotlinType.constructorName
     get() = constructor.declarationDescriptor?.fqNameSafe?.asString()
 
+sealed class DriTarget {
+    override fun toString(): String = this.javaClass.simpleName
+
+    companion object {
+        fun from(descriptor: DeclarationDescriptor): DriTarget = descriptor.parentsWithSelf.run {
+            val callable = firstIsInstanceOrNull<CallableDescriptor>()
+            val params = callable?.let { listOfNotNull(it.extensionReceiverParameter) + it.valueParameters }.orEmpty()
+            val parameterDescriptor = firstIsInstanceOrNull<ParameterDescriptor>()
+
+            return when(descriptor){
+                is TypeParameterDescriptor -> PointingToGenericParameters(descriptor.index)
+                is ReceiverParameterDescriptor -> PointingToReceiver
+                is ClassDescriptor, is FunctionDescriptor -> PointingToFunctionOrClasslike
+                is PropertyDescriptor -> PointingToProperty
+                else -> parameterDescriptor?.let { PointingToAdditionalParameters(params.indexOf(it)) }
+                    ?: PointingToOther(descriptor.name.toString())
+            }
+        }
+
+        fun from(psi: PsiElement): DriTarget = psi.parentsWithSelf.run {
+            val callable = firstIsInstanceOrNull<PsiMethod>()
+            val params = (callable?.parameterList?.parameters).orEmpty()
+            firstIsInstanceOrNull<PsiParameter>()?.let { params.indexOf(it) }
+
+            return when(psi) {
+                is PsiTypeParameter -> PointingToGenericParameters(psi.index)
+                is PsiReceiverParameter -> PointingToReceiver
+                is PsiClass, is PsiMethod -> PointingToFunctionOrClasslike
+                is PsiField -> PointingToProperty
+                else -> firstIsInstanceOrNull<PsiParameter>()?.let { PointingToAdditionalParameters(params.indexOf(it)) }
+                    ?: PointingToOther(psi.text)
+            }
+        }
+    }
+}
+
+data class PointingToGenericParameters(val parameterIndex: Int) : DriTarget() {
+    override fun toString(): String = "PointingToGenericParameters($parameterIndex)"
+}
+
+object PointingToFunctionOrClasslike: DriTarget()
```

Comment:
Pointing to a declaration is not the best solution, since everything that is passed to a DRI constructing function is a declaration. 

I've split this to make it more readable from the perspective of somebody reading those DRIs.